src/app/functions/server/getRelevantGuidedSteps.ts

```diff
@@ -0,0 +1,260 @@
+  const stepLinksToCheck =
+    fromIndex !== -1 ? stepLinkStatuses.slice(fromIndex) : stepLinkStatuses;
+  const nextStep = stepLinksToCheck.find((stepLink) => {
+    return stepLink.eligible && !stepLink.completed;
+  });
+
+  // There should be a relevant next step for every user (even if it's just
+  // going back to the dashbord), so we can't hit this line in tests (and
+  // shouldn't be able to in production either):
+  /* c8 ignore next */
+  return nextStep ?? null;
+}
+
+export function getGuidedStepStatuses(data: InputData): StepLinkWithStatus[] {
+  return stepLinks.map((stepLink) => getStepWithStatus(data, stepLink));
+}
+
+function getStepWithStatus(
+  data: InputData,
+  stepLink: StepLink
+): StepLinkWithStatus {
+  return {
+    ...stepLink,
+    eligible: isEligibleFor(data, stepLink.id),
+    completed: hasCompleted(data, stepLink.id),
+  };
+}
+
+function isEligibleFor(data: InputData, stepId: StepLink["id"]): boolean {
+  if (stepId === "StartScan") {
+    return data.countryCode === "us";
+  }
+
+  if (stepId === "ScanResult") {
+    return data.countryCode === "us";
+  }
+
+  if (stepId === "HighRiskSsn") {
+    // Our social security number-related mitigations aren't possible outside of the US:
+    return data.countryCode === "us";
+  }
+
+  if (
+    ["HighRiskCreditCard", "HighRiskBankAccount", "HighRiskPin"].includes(
+      stepId
+    )
+  ) {
+    // Anyone can view/resolve their high risk data breaches
+    return true;
+  }
+
+  if (
+    ["LeakedPasswordsPassword", "LeakedPasswordsSecurityQuestion"].includes(
+      stepId
+    )
+  ) {
+    // Anyone can view/resolve their leaked passwords
+    return true;
+  }
+
+  if (
+    ["SecurityTipsPhone", "SecurityTipsEmail", "SecurityTipsIp"].includes(
+      stepId
+    )
+  ) {
+    // Anyone can view security tips
+    return true;
+  }
+
+  if (stepId === "Done") {
+    return true;
+    // All steps should have been covered by the above conditions:
+    /* c8 ignore next 4 */
+  }
+
+  return false as never;
+}
+
+function hasCompleted(data: InputData, stepId: StepLink["id"]): boolean {
+  if (stepId === "StartScan") {
+    return data.latestScanData?.scan !== null;
+  }
+
+  if (stepId === "ScanResult") {
+    return (
+      data.latestScanData !== null &&
+      Array.isArray(data.latestScanData?.results) &&
+      !data.latestScanData.results.some((scanResult) => {
+        return scanResult.status === "new" && !scanResult.manually_resolved;
+      })
+    );
+  }
+
+  function isResolved(
+    dataClass: (typeof BreachDataTypes)[keyof typeof BreachDataTypes]
+  ): boolean {
+    return !data.subscriberBreaches.some((breach) => {
+      return (
+        breach.dataClasses.includes(dataClass) &&
+        !breach.resolvedDataClasses.includes(dataClass)
+      );
+    });
+  }
+
+  // TODO: instead of resolving entire breaches, only resolve specific data classes
+  //       See ADR 0008-data-class-resolution.
```

Comment:
I was not able to find the ADR. Didn’t we already find out that we might not need to restructure in order to be able to resolve specific data classes?

Reply:
Whoops. I actually have one written out, but then I found out that we don't actually need to resolve entire breaches (see `isResolved` above, which _does_ look at the specific data class), and so it was no longer needed. I just forgot to remove this comment. Removed in 74b759d1d.